### DISCUSSION TITLE
Null pointer in NetLicensing client code

### DISCUSSION THREAD

**Discussion by DimaWhite** [Labs64/NetLicensing-Community#67] [General]
I tried creating a test setup, and get an null exception!  (Not a great programming experience)  My code is as follows:

                ValidationParameters validationParameters = new ValidationParameters();
                validationParameters.setLicenseeName(demoLicenseeName);
                validationParameters.setProductNumber(demoProductNumber);
                validationParameters.setLicenseeSecret(demoLicenseeSecret);
                validationParameters.put(demoProductModuleNumber, "paramKey", "paramValue");
                context.securityMode = SecutiryMode.APIKEY_IDENTIFICATION;
                ValidationResult validationResult = LicenseeService.validate(context, demoLicenseeNumber, validationParameters);

Can I created everything in the account...  Can you please take a look and tell me why this occurred?

Thanks!

**Comment by kkorotkov** [comment]
Hello @DimaWhite,

Is there any chance you could share entire Java Class, so we can see the implementation details.
As well as the error stacktrace.
We assume, that the message in the source code itself, as NetLicensing is not returning null exceptions, rather corresponding HTTP status code.

**Comment by DimaWhite** [comment]
Sure, It's just a modified version of your example/unit test code:
namespace NetLicensingClient
{
    class NetLicensingClient_demo
    {
        static int Main(string[] args)
        {
            // ServicePointManager.ServerCertificateValidationCallback = delegate { return true;  // Trust any (self-signed) certificate };

            Context context = new Context();
            context.baseUrl = "https://go.netlicensing.io";
            context.apiKey = "221e2277-9fc8-8b20-32ed-13738dda919c";
            context.securityMode = SecutiryMode.APIKEY_IDENTIFICATION;

	    String randomNumber = randomString(8);

            String demoProductNumber = numberWithPrefix("P", randomNumber);
            String demoProductModuleNumber = numberWithPrefix("PM", randomNumber);
            String demoLicensingModel = Constants.LicensingModel.TryAndBuy.NAME;
            String demoLicenseTemplate1_Number = numberWithPrefix("LT", randomNumber);
            String demoLicenseTemplate1_Name = "Demo Evaluation Period";
            String demoLicenseTemplate1_Type = "FEATURE";
            Decimal demoLicenseTemplate1_Price = 0M;
            String demoLicenseTemplate1_Currency = "USD";
            Boolean demoLicenseTemplate1_Automatic = false;
            Boolean demoLicenseTemplate1_Hidden = false;
            String demoLicenseeNumber = numberWithPrefix("L", randomNumber);
            String demoLicenseNumber = numberWithPrefix("LC", randomNumber);
            String demoLicenseeName = "Demo Licensee";
            String demoLicenseeSecret = "221e22779fc88b2032ed13738dda919c";
            String demoProductLicenseeSecretMode = "PREDEFINED";

            try
            {
                //#region ****** Lists

                //List<String> licenseTypes = UtilityService.listLicenseTypes(context);
                //ConsoleWriter.WriteList("License Types:", licenseTypes);

                //List<String> licensingModels = UtilityService.listLicensingModels(context);
                //ConsoleWriter.WriteList("Licensing Models:", licensingModels);

                //#endregion

                //#region ****** Product

                //Product product = ProductService.get(context, demoProductNumber);
                //ConsoleWriter.WriteEntity("Got product:", product);

                //List<Product> products = ProductService.list(context, null);
                //ConsoleWriter.WriteList("Got the following products:", products);

                //#endregion

                //#region ****** Transaction

                //List<Transaction> transactions = TransactionService.list(context, null);
                //ConsoleWriter.WriteList("Got the following transactions:", transactions);

                //#endregion

                //#region ****** ProductModule

                //ProductModule productModule = ProductModuleService.get(context, demoProductModuleNumber);
                //ConsoleWriter.WriteEntity("Got product module:", productModule);

                //List<ProductModule> productModules = ProductModuleService.list(context, null);
                //ConsoleWriter.WriteList("Got the following ProductModules:", productModules);

                //#endregion

                //#region ****** LicenseTemplate
                //LicenseTemplate licenseTemplate = LicenseTemplateService.get(context, demoLicenseTemplate1_Number);
                //ConsoleWriter.WriteEntity("Got licenseTemplate:", licenseTemplate);

                //List<LicenseTemplate> licenseTemplates = LicenseTemplateService.list(context, null);
                //ConsoleWriter.WriteList("Got the following license templates:", licenseTemplates);

                //#endregion

                //#region ****** Licensee

                //List<Licensee> licensees = LicenseeService.list(context, null);
                //ConsoleWriter.WriteList("Got the following licensees:", licensees);

                //#endregion

                //#region ****** License
                //List<License> licenses = LicenseService.list(context, null);
                //ConsoleWriter.WriteList("Got the following licenses:", licenses);

                //#endregion

                //#region ****** PaymentMethod

                //List<PaymentMethod> paymentMethods = PaymentMethodService.list(context);
                //ConsoleWriter.WriteList("Got the following payment methods:", paymentMethods);

                //#endregion

                //#region ****** Validate

                ValidationParameters validationParameters = new ValidationParameters();
                validationParameters.setLicenseeName(demoLicenseeName);
                validationParameters.setProductNumber(demoProductNumber);
                validationParameters.setLicenseeSecret(demoLicenseeSecret);
                validationParameters.put(demoProductModuleNumber, "paramKey", "paramValue");
                context.securityMode = SecutiryMode.APIKEY_IDENTIFICATION;
                ValidationResult validationResult = LicenseeService.validate(context, demoLicenseeNumber, validationParameters);
                ConsoleWriter.WriteEntity("Validation repeated with APIKey:", validationResult);
                Console.WriteLine("All done.");

                return 0;
            }
            catch (NetLicensingException e)
            {
                Console.WriteLine("Got NetLicensing exception:");
                Console.WriteLine(e);
            }
            catch (Exception e)
            {
                Console.WriteLine("Got exception:");
                Console.WriteLine(e);
            }

            return 1;
        }

	public static string randomString(int length)
	{
	    Random random = new Random();
	    const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
	    return new string(Enumerable.Repeat(chars, length)
	      .Select(s => s[random.Next(s.Length)]).ToArray());
	}

	private static String numberWithPrefix(String prefix, String number)
	{
	    return "DEMO-"+prefix+number;
	}
	                
    }
}

**Comment by kkorotkov** [comment]
Aha, it's C# and not Java. So, there are several points:

- First of all, I see you're using very old code. It was once published by one of our C# early adopters, but it was unlisted already since quite a while. I suggest to switch to our officially supported and maintained client library that you can find here: https://github.com/Labs64/NetLicensingClient-csharp

- Next, I see in this code many calls to create entities are commented-out, and only `validate` remains. At the same time entity reference numbers are randomly generated. `validate` requires entities to be present, which can't be fulfilled with the randomly generated numbers, and no code to create entities with these numbers. Here I suggest the following approach for the proof-of-concept:
  1. In your account, generate test entities for the desired licensing model using "Demo Data" option in the NetLicensing UI (user menu in the top-right corner);
  2. Use the numbers of the generated entities in your code;
  3. Once ready for production implementation, you can replace demo data with the real data.

- Third, you're trying to use a deprecated "Licensee Secret" functionality. This is discouraged, please use "Node-Locked" licensing model instead - it fully covers functionality of the "Licensee Secret" and adds more functionality, like multiple secrets per licensee.

Let us know if above suggestions solve the problem.

**Comment by DimaWhite** [comment]
Evaluated with the latest client library - looks much better, so the issue is solved.
